src/com/b44t/messenger/rpc/Rpc.java

```diff
@@ -0,0 +1,120 @@
+package com.b44t.messenger.rpc;
+
+import com.b44t.messenger.DcJsonrpcInstance;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonSyntaxException;
+import com.google.gson.reflect.TypeToken;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+
+public class Rpc {
+    private Map<Integer, CompletableFuture<JsonElement>> requestFutures = new ConcurrentHashMap<>();
+    private DcJsonrpcInstance dcJsonrpcInstance;
+    private int requestId = 0;
+    private Gson gson = new GsonBuilder().serializeNulls().create();
+
+    public Rpc(DcJsonrpcInstance dcJsonrpcInstance) {
+        this.dcJsonrpcInstance = dcJsonrpcInstance;
+    }
+
+    private void processResponse() throws JsonSyntaxException {
+        String jsonResponse = dcJsonrpcInstance.getNextResponse();
+        Response response = gson.fromJson(jsonResponse, Response.class);
+
+        if (response.id == 0) { // Got JSON-RPC notification, ignore
+            return;
+        }
+
+        CompletableFuture<JsonElement> future = requestFutures.remove(response.id);
+        if (response.error != null) {
+            future.completeExceptionally(new RpcError(response.error.toString()));
+        } else if (response.result != null) {
+            future.complete(response.result);
+        } else {
+            future.completeExceptionally(new RpcError("Got JSON-RPC response witout result or error: " + jsonResponse));
+        }
```

Comment:
There's a lot of code changing exceptions from one form to another here, what if we replaced this whole block with `future.complete(response);`, used `CompletableFuture<Response>` instead of `CompletableFuture<JsonElement>`, and let the callers (like `getResult()`) extract the error on their own.

Also, creating a Throwable is always expensive, since it captures a stack trace. Idk how expensive excactly and if it might be noticeable in our code.

Reply:
all the callers will want to extract the error/result so it would be too much of a burden and repeated code, they have to do this anyway, and if there is an error they will want to print it in the logs, because it will be a programming error (ex. bad usage of API)

Reply:
(this are not incoming events that could become a bunch of errors, they are only responses to API calls, so if you call a function and it throws an exception you need to handle it anyway)

Reply:
But checking for response.result != null is the same amount of code as catching the error